File: deepeval/metrics/dag/runner.py
from __future__ import annotations

import asyncio
from typing import TYPE_CHECKING, Any, Dict, List, Optional, Tuple, Union

from deepeval.metrics.dag.nodes import (
    BaseNode,
    VerdictNode,
    TaskNode,
    BinaryJudgementNode,
    NonBinaryJudgementNode,
    construct_node_verbose_log as _st_verbose_log,
)
from deepeval.metrics.conversational_dag.nodes import (
    ConversationalBaseNode,
    ConversationalVerdictNode,
    ConversationalTaskNode,
    ConversationalBinaryJudgementNode,
    ConversationalNonBinaryJudgementNode,
    construct_node_verbose_log as _mt_verbose_log,
)
from deepeval.metrics.base_metric import BaseMetric, BaseConversationalMetric
from deepeval.test_case import LLMTestCase, ConversationalTestCase

if TYPE_CHECKING:
    from deepeval.metrics.dag.graph import DeepAcyclicGraph

Node = Union[BaseNode, ConversationalBaseNode]
Metric = Union[BaseMetric, BaseConversationalMetric]
TestCase = Union[LLMTestCase, ConversationalTestCase]

_NODE = (BaseNode, ConversationalBaseNode)
_VERDICT = (VerdictNode, ConversationalVerdictNode)
_TASK = (TaskNode, ConversationalTaskNode)
_JUDGEMENT = (
    BinaryJudgementNode,
    NonBinaryJudgementNode,
    ConversationalBinaryJudgementNode,
    ConversationalNonBinaryJudgementNode,
)


class _DeepAcyclicGraphRunner:
    def __init__(self, graph: DeepAcyclicGraph):
        self.graph = graph
        self.remaining: Dict[Node, int] = dict(graph.indegree)
        self.outputs: Dict[Node, Any] = {}
        self.verdicts: Dict[Node, Any] = {}
        self.depth: Dict[Node, int] = {}
        # Every fired scoring verdict/child metric contributes a (score, reason)
        # pair; the final metric score is resolved deterministically at the end
        # instead of "last write wins" (which depended on node order and on the
        # sync vs async traversal).
        self.scores: List[Tuple[float, Optional[str]]] = []
        self._verbose_log = (
            _mt_verbose_log if graph.multiturn else _st_verbose_log
        )

    def _check_remaining(self, node: Node) -> bool:
        self.remaining[node] -= 1
        return self.remaining[node] <= 0

    def _verdict_matches(self, node: Node) -> bool:
        parents = self.graph.parents.get(node)
        parent = parents[0] if parents else None
        if isinstance(parent, _JUDGEMENT):
            return self.verdicts[parent].verdict == node.verdict
        return True

    def _store_result(self, node: Node, result: Any) -> None:
        if isinstance(node, _TASK):
            self.outputs[node] = result
        else:
            self.verdicts[node] = result

    def _apply_child_metric(
        self, node: Node, child_metric: Metric, metric: Metric, depth: int
    ) -> None:
        metric._verbose_steps.append(
            self._verbose_log(node, depth, child_metric)
        )
        metric.score = child_metric.score
        reason = child_metric.reason if metric.include_reason else None
        self.scores.append((child_metric.score, reason))
        metric._accrue_cost(child_metric.evaluation_cost)
        metric._accrue_tokens(
            child_metric.input_tokens, child_metric.output_tokens
        )

    def _finalize_score(self, metric: Metric) -> None:
        """Resolve the final metric score deterministically.

        Every fired scoring verdict/child metric contributes a (score, reason)
        pair. The score is picked with a conservative ``min`` aggregation so
        the result no longer depends on node declaration order or on whether
        the graph ran in sync or async mode. A DAG with a single scoring
        path is unaffected.
        """
        scored = [(s, r) for s, r in self.scores if s is not None]
        if not scored:
            return
        score, reason = min(scored, key=lambda item: item[0])
        metric.score = score
        if metric.include_reason and reason is not None:
            metric.reason = reason

    # ------------------------------------------------------------------ sync
    def run(self, metric: Metric, test_case: TestCase) -> None:
        for root in self.graph.root_nodes:
            self._visit(root, metric, test_case, 0)
        self._finalize_score(metric)

    def _visit(
        self, node: Node, metric: Metric, test_case: TestCase, depth: int
    ) -> None:
        if isinstance(node, _VERDICT):
            self._visit_verdict(node, metric, test_case, depth)
            return
        self.depth[node] = max(0, self.depth.get(node, 0), depth)
        if not self._check_remaining(node):
            return
        result = node._execute(
            metric, test_case, self.graph.parents.get(node), self.outputs
        )
        self._store_result(node, result)
        node_depth = self.depth[node]
        metric._verbose_steps.append(
            self._verbose_log(
                node,
                node_depth,
                output=self.outputs.get(node),
                verdict=self.verdicts.get(node),
            )
        )
        for child in node.children:
            self._visit(child, metric, test_case, node_depth + 1)

    def _visit_verdict(
        self, node: Node, metric: Metric, test_case: TestCase, depth: int
    ) -> None:
        if not self._check_remaining(node):
            return
        if not self._verdict_matches(node):
            return
        child = node.child
        if child is None:
            metric._verbose_steps.append(self._verbose_log(node, depth))
            metric.score = node.score / 10
            reason = (
                node._generate_reason(metric=metric)
                if metric.include_reason
                else None
            )
            self.scores.append((metric.score, reason))
        elif isinstance(child, _NODE):
            self._visit(child, metric, test_case, depth)
        else:
            copied = node._run_child_metric(metric, test_case)
            self._apply_child_metric(node, copied, metric, depth)

    async def a_run(self, metric: Metric, test_case: TestCase) -> None:
        await asyncio.gather(
            *(
                self._a_visit(root, metric, test_case, 0)
                for root in self.graph.root_nodes
            )
        )
        self._finalize_score(metric)

    async def _a_visit(
        self, node: Node, metric: Metric, test_case: TestCase, depth: int
    ) -> None:
        if isinstance(node, _VERDICT):
            await self._a_visit_verdict(node, metric, test_case, depth)
            return
        self.depth[node] = max(0, self.depth.get(node, 0), depth)
        if not self._check_remaining(node):
            return
        result = await node._a_execute(
            metric, test_case, self.graph.parents.get(node), self.outputs
        )
        self._store_result(node, result)
        node_depth = self.depth[node]
        metric._verbose_steps.append(
            self._verbose_log(
                node,
                node_depth,
                output=self.outputs.get(node),
                verdict=self.verdicts.get(node),
            )
        )
        await asyncio.gather(
            *(
                self._a_visit(child, metric, test_case, node_depth + 1)
                for child in node.children
            )
        )

    async def _a_visit_verdict(
        self, node: Node, metric: Metric, test_case: TestCase, depth: int
    ) -> None:
        if not self._check_remaining(node):
            return
        if not self._verdict_matches(node):
            return
        child = node.child
        if child is None:
            metric._verbose_steps.append(self._verbose_log(node, depth))
            metric.score = node.score / 10
            reason = (
                await node._a_generate_reason(metric=metric)
                if metric.include_reason
                else None
            )
            self.scores.append((metric.score, reason))
        elif isinstance(child, _NODE):
            await self._a_visit(child, metric, test_case, depth)
        else:
            copied = await node._a_run_child_metric(metric, test_case)
            self._apply_child_metric(node, copied, metric, depth)

File: typescript/src/metrics/dag/runner.ts
import type { DeepAcyclicGraph } from "@/metrics/dag/graph";
import type { JudgementVerdict } from "@/metrics/dag/schema";
import {
  type AnyNode,
  type ChildMetric,
  type DagHostMetric,
  type ExecutableNode,
  type VerdictLikeNode,
  isJudgementNode,
  isNode,
  isVerdictNode,
} from "@/metrics/dag/types";

export class DeepAcyclicGraphRunner {
  private readonly remaining: Map<AnyNode, number>;
  private readonly outputs = new Map<AnyNode, unknown>();
  private readonly verdicts = new Map<AnyNode, JudgementVerdict>();
  private readonly depth = new Map<AnyNode, number>();

  constructor(private readonly graph: DeepAcyclicGraph) {
    this.remaining = new Map(graph.indegree);
  }

  async run(metric: DagHostMetric, testCase: any): Promise<void> {
    await Promise.all(
      this.graph.rootNodes.map((root) => this.visit(root, metric, testCase, 0)),
    );
  }

  private checkRemaining(node: AnyNode): boolean {
    const left = (this.remaining.get(node) ?? 0) - 1;
    this.remaining.set(node, left);
    return left <= 0;
  }

  private verdictMatches(node: VerdictLikeNode): boolean {
    const parent = this.graph.parents.get(node)?.[0];
    if (parent && isJudgementNode(parent)) {
      return this.verdicts.get(parent)?.verdict === node.verdict;
    }
    return true;
  }

  private storeResult(node: ExecutableNode, result: unknown): void {
    if (node.nodeKind === "task") {
      this.outputs.set(node, result);
    } else {
      this.verdicts.set(node, result as JudgementVerdict);
    }
  }

  private async visit(
    node: AnyNode,
    metric: DagHostMetric,
    testCase: any,
    depth: number,
  ): Promise<void> {
    if (isVerdictNode(node)) {
      await this.visitVerdict(node, metric, testCase, depth);
      return;
    }
    this.depth.set(node, Math.max(0, this.depth.get(node) ?? 0, depth));
    if (!this.checkRemaining(node)) return;

    const result = await node.execute({
      metric,
      testCase,
      parents: this.graph.parents.get(node),
      outputs: this.outputs,
    });
    this.storeResult(node, result);

    const nodeDepth = this.depth.get(node)!;
    metric.verboseSteps.push(node.verboseLog(nodeDepth, result));
    await Promise.all(
      node.children.map((child) =>
        this.visit(child, metric, testCase, nodeDepth + 1),
      ),
    );
  }

  private async visitVerdict(
    node: VerdictLikeNode,
    metric: DagHostMetric,
    testCase: any,
    depth: number,
  ): Promise<void> {
    if (!this.checkRemaining(node)) return;
    if (!this.verdictMatches(node)) return;

    const child = node.child;
    if (child == null) {
      metric.verboseSteps.push(node.verboseLog(depth));
      metric.score = node.score! / 10;
      if (metric.includeReason) {
        metric.reason = await node.generateReason(metric);
      }
    } else if (isNode(child)) {
      await this.visit(child, metric, testCase, depth);
    } else {
      const childMetric = await node.runChildMetric(metric, testCase);
      this.applyChildMetric(node, childMetric, metric, depth);
    }
  }

  private applyChildMetric(
    node: VerdictLikeNode,
    childMetric: ChildMetric,
    metric: DagHostMetric,
    depth: number,
  ): void {
    metric.verboseSteps.push(node.verboseLog(depth, childMetric));
    metric.score = childMetric.score;
    if (metric.includeReason) metric.reason = childMetric.reason;
    metric.accrueCost(childMetric.evaluationCost ?? null);
  }
}
